### ISSUE TITLE
Proposal: "implies" logical operator

### ISSUE THREAD

**Issue by sjb-sjb** [dotnet/csharplang#1439]
It would be great if we had an operator for logical implication, namely "a implies b" is equivalent to "!a || b". The lazy evaluation would also be similar, i.e. b is only evaluated if !a is true. 

Why? Because it is needed all the time for assertions, and it is too hard to see the meaning of "!a || b" when you are trying to express implication. 

As for the symbol to use for this, how about "=>" ? There are other possibilities such as "->" or ">>", but overall I think "=>" would be the best. Note it cannot be confused with the other uses of "=>".

**Comment by jnm2** [comment]
> Note it cannot be confused with the other uses of "=>".

With expression bodies and lambdas everywhere, I would find this very hard to parse visually and I'm pretty sure folks will come up with plenty of compiler ambiguity issues.

If we had `imp`, for the same of argument because I find that easier to wrap my head around right now, it doesn't read properly in my head:

```cs
if (foo.Equals(42) imp bar.IsSet) // Reads as "if foo equaling 42 implies bar is set"
{
    // ...
}

if (!foo.Equals(42) || bar.IsSet) // I'm used to this; I get it immediately
{
    // ...
}
```

I'm lost in the first example. It sounds like I'm talking in terms of a generalization about the source code, not in terms of actually evaluating the operands.

**Comment by sjb-sjb** [comment]
OK so some examples: 

Debug.Assert( this.EntityIsToBeDeleted => (this.Id >= 0) ); // asserting relationship between state properties
Debug.Assert( (serviceObject != null) => (serviceObject.Capacity >= 1)); // asserting that if a service object is provided then it must have sufficient capacity. 
Debug.Assert( (this.IsExecuting) => (this.RemainingCount > 0) );  // asserting relationship between state properties

Or for example, 
bool isValid = Validator.TryValidateObject( this, validationContext, validationMessages)
Debug.Assert( !isValid => (validationMessages.Count > 0)); // asserting that if validation failed then specific error messages were produced.

P.S. Another implementation would be:  a? b: true. The only thing is that if "b" is really long syntactically then you can't figure out what is being said until you get to the very end of the line where it says "true".

**Comment by jnm2** [comment]
Are you envisioning that this would typically be used with debug assertions rather than `if` statements?

**Comment by bondsbw** [comment]
I might say to my sophisticated toddler, "eating your vegetables implies that you can have dessert."

So how does that fit in an `if` block?

```c#
if (you.EatVegetables implies you.CanHaveDessert)
{
    // What makes sense here? 
}
```

I think it would make more sense as a definition instead of imperative:

```c#
public property EatVegetables
{
    implies CanHaveDessert;
}
```

Or in the original proposal syntax:

```c#
public property EatVegetables => CanHaveDessert;
```

Which, of course, is already possible.

**Comment by ufcpp** [comment]
Are you unsatisfied with extension methods?

```cs
static class BoolExtensions
{
    public static bool Implies(this bool a, bool b) => !a || b;
}
```

```cs
Debug.Assert( this.EntityIsToBeDeleted.Implies(this.Id >= 0)); 
```

**Comment by Starwer** [comment]
Probably a stupid question... How would is the hypothetical `imply` operator different than the `&&` operator ?

```C#
if (you.EatVegetables && you.CanHaveDessert)
{
    // no ambiguity here !
}

bool isValid = Validator.TryValidateObject( this, validationContext, validationMessages)
Debug.Assert( !isValid && (validationMessages.Count > 0)); // asserting that if validation failed then specific error messages were produced.

```

**Comment by Starwer** [comment]
Ok... that was a stupid question... I think I got it now how it could be used in assertions. Sorry for this.

**Comment by IanWold** [comment]
I would actually be able to make use of this a good bit, but I do not find `(!a||b)` to be cumbersome nor difficult to get my head around. Is there another reason this might be necessary?

**Comment by CyrusNajmabadi** [comment]
> Are you unsatisfied with extension methods?

While i don't actually really care about an implies-operator, i will note that an extension method would not suffice.  This is because it would not actually have the short-circuiting behavior one might want/expect of such an operator.

With your example: ```this.EntityIsToBeDeleted.Implies(this.Id >= 0)```, say that "id" throws if "this.EntityIsToBeDeleted" is false.  With a shortcircuiting operator, you're fine.  But with hte extension method, you'll throw as "this.Id >= 0" is always executed.

**Comment by jnm2** [comment]
@CyrusNajmabadi If `this.Id >= 0` threw an exception, that would be a better example. 😄

**Comment by jnm2** [comment]
@CyrusNajmabadi If this.Id >= 0 threw an exception, that would be a better example. 😄

How about  `(entity != null).Implies(entity.Id >= 0)`?

**Comment by jnm2** [comment]
Oh, duh...

**Comment by Starwer** [comment]
FYI, this operator indeed exists in Hardware Design Language System Verilog, for assertions indeed, and it is: `|->`. I personally found it "nice" but unnecessary. Especially when you think about the  target audience: an HDL designer uses Boolean arithmetic more than the numeric addition.

**Comment by sjb-sjb** [comment]
@CyrusNajmabadi has a good point. If we extend the ordering operators ( >=, <=, >, < ) to operate on Boolean values then we are committing to evaluating both arguments. On the other hand if we create a new operator " => " for implication then we are able to have conditional / partial evaluation.

**Comment by Neme12** [comment]
I find it hard to even understand what this means. Can you show a use case for this?

**Comment by sjb-sjb** [comment]
OK, then, as originally suggested: a new Boolean operator a => b which is equivalent to !a | b. 

This is not meant to be a big idea, this is a small idea !!  This is a $0.02 suggestion. 

@Neme12 for use cases check out the earlier discussion.

**Comment by HaloFour** [comment]
@sjb-sjb 

The syntax `a => b` is already a valid expression.  It defines a lambda.

**Comment by sjb-sjb** [comment]
Right ... I guess if "a" is a variable rather than some other expression then this is ambiguous syntax. We just need some other symbol. Personally I like the ordering operators >= etc, notwithstanding @CyrusNajmabadi 's  comment and the question about partialness. However if we are going to stick with the partial valuation approach then how about "->"  ?  It does look a little funny if you are used to C or C++.

**Comment by Neme12** [comment]
@sjb-sjb 
>  It does look a little funny if you are used to C or C++.

Or C#. Recall that it has pointers as well:
```c#
using System;

class C
{
    unsafe static void Main()
    {
        var d = DateTime.Today;
        DateTime* p = &d;
        Console.WriteLine(p->Date);
    }
}
```

**Comment by Starwer** [comment]
@ufcpp  & @jnm2 : That's right, we can do it with some clever extension methods. However, I don't see how these extension methods help in making thing more readable than a simple `!a || b` expression. I guess there are many different ways to do such an imply feature. But the question of @sjb-sjb  was how to make it look nicer I think... I also believe the Perl motto: "There's more than one way to do it" doesn't fit the C# Philosophy so well.

**Comment by jnm2** [comment]
@Starwer I didn't suggest using an extension method. I agree with you.
